substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -119,84 +419,1339 @@ protected int peekData(DatagramPacket p) throws IOException {
+                            } else if (ret == Target_jvm.JVM_IO_ERR()) {
+                                //   853                       if (errno == ENOMEM) {
+                                if (Errno.errno() == Errno.ENOMEM()) {
+                                    //   854                          JNU_ThrowOutOfMemoryError(env, "NET_Timeout native heap allocation failed");
+                                    throw new OutOfMemoryError("NET_Timeout native heap allocation failed");
+                                    //   855  #ifdef __linux__
+                                } else if (IsDefined.__linux__()) {
+                                    //   856                       } else if (errno == EBADF) {
+                                    if (Errno.errno() == Errno.EBADF()) {
+                                        //   857                           JNU_ThrowByName(env, JNU_JAVANETPKG "SocketException", "Socket closed");
+                                        throw new SocketException("Socket closed");
+                                    } else {
+                                        //   859                           NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException", "Receive failed");
+                                        throw new SocketException("Receive failed");
+                                    }
+                                    //   860  #else
+                                } else {
+                                    //   862                           JNU_ThrowByName(env, JNU_JAVANETPKG "SocketException", "Socket closed");
+                                    throw new SocketException("Socket closed");
+                                    //   863  #endif
+                                }
+
+                                //   865                  } else if (ret == JVM_IO_INTR) {
+                            } else if (ret == Target_jvm.JVM_IO_INTR()) {
+                                //   866                      JNU_ThrowByName(env, JNU_JAVAIOPKG "InterruptedIOException",
+                                //   867                                      "operation interrupted");
+                                throw new InterruptedIOException("operation interrupted");
+                            }
+                        } finally {
+                            //   870                  if (mallocedPacket) {
+                            if (mallocedPacket) {
+                                //   871                      free(fullPacket);
+                                LibC.free(fullPacket);
+                                /* so that the outter try/finally does not attempt to double-free() */
+                                mallocedPacket = false;
+                            }
+                            //   874                  return;
+                            return;
+                        }
```

Comment:
`javac` complains on this line
```
graal/substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java:609: warning: [finally] finally clause cannot complete normally
                        }
                        ^
```
which is not much help until you look up what the error message means: that you have a `return` in a `finally` block.

I think you have to either put the `return` just after of this `try/finally`, or trust the outer `try/finally` to free `fullPacket` when (not if!) an exception gets thrown from this code.  (Your translation is already aware of the "outter" (sic) `try/finally`.)  I would be in favor of one `try/finally` starting right after the potential `malloc` and enforcing the potential `free`.

Reply:
I get that same warning on lines 609, 996, 1024, 1040, 1060, 1086, and 1117.  I am surprised you don't get them.  We do not allow code that generates warnings.